### PR TITLE
[prometheus] exporter prototype

### PR DESCRIPTION
**Pull request by mikz**
* introduce metrics phase to export metrics
* executor prints collects the response from prometheus
* metrics can be also published from any other phase or timer

## Example
See for policy changes and metrics examples:
https://github.com/3scale/apicast-cloud-hosted/pull/5

### CODE REVIEW

File: gateway/config/development.lua [hunk 1]
```diff
@@ -28,5 +28,6 @@ return {
     lua_code_cache = 'on',
     configuration_loader = 'lazy',
     configuration_cache = 0,
-    configuration = data_url('application/json', configuration)
+    configuration = data_url('application/json', configuration),
+    port = { metrics = 9100 },
```
Comment (maneta) [inline gateway/config/development.lua]: @mikz maybe reserve a port for the APIcast endpoint so we can publish it on the [Prometheus Default Port Allocation list](https://github.com/prometheus/prometheus/wiki/Default-port-allocations)?

Comment (mikz) [reply gateway/config/development.lua]: I think we should agree how we want to proceed first.
We are not going to reserve a port for every app we have, right? We need some generic port for openresty exporter?

Comment (maneta) [reply gateway/config/development.lua]: Got It, I thought that at least for APICast we were willing to create a specific "exporter" but yes makes more sense to create a generic port for openresty exporter.

Comment (jmprusi) [reply gateway/config/development.lua]: To me it doesn't really make sense to reserve a port for each APP, but this is not a /generic/ openresty prometheus exporter, so, should we define a port for the "APIcast framework" and reuse that one in all our applications?

Comment (mikz) [reply gateway/config/development.lua]: That makes sense if we want to converge all our proxies running on the "APIcast framework".

File: CHANGELOG.md [hunk 2]
```diff
@@ -11,6 +11,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - Error loading policy chain configuration JSON with null value [PR #626](https://github.com/3scale/apicast/pull/626)
 - Splitted `resolv.conf` in lines,to avoid commented lines  [PR #618](https://github.com/3scale/apicast/pull/618)
 
+## Addwed
```
Comment (davidor) [inline CHANGELOG.md]: Typo

File: gateway/config/development.lua (hunk 1 above)
Comment (mikz) [reply gateway/config/development.lua]: Reserved 9421.

Comment (jmprusi) [reply gateway/config/development.lua]: We should converge them all, now with policies it should be easier ;)

File: gateway/conf/nginx.conf.liquid [hunk 3]
```diff
@@ -106,6 +106,24 @@ http {
     {% include "conf.d/apicast.conf" %}
   }
 
+  {% if port.metrics %}
+    lua_shared_dict prometheus_metrics 16M;
```
Comment (davidor) [inline gateway/conf/nginx.conf.liquid]: Should this be configurable?

Comment (mikz) [reply gateway/conf/nginx.conf.liquid]: Probably, but I'd wait with some real operations use to see if it is really needed.

File: gateway/conf/nginx.conf.liquid [hunk 4]
```diff
@@ -106,6 +106,24 @@ http {
...
+    server {
+      access_log off;
+      listen {{ port.metrics }};
+      server_name metrics prometheus _;
+
+      location /metrics {
+        content_by_lua_block { require('apicast.executor'):metrics() }
+      }
+
+      location /nginx_status {
```
Comment (davidor) [inline gateway/conf/nginx.conf.liquid]: Why is this needed?

File: gateway/conf/nginx.conf.liquid (hunk 3 above)
Comment (mikz) [reply gateway/conf/nginx.conf.liquid]: Hah. This just gave me an idea for another exporter: shmem dict stats about free space.

File: gateway/config/development.lua [hunk 5]
```diff
@@ -28,5 +28,6 @@ return {
     lua_code_cache = 'on',
     configuration_loader = 'lazy',
     configuration_cache = 0,
-    configuration = data_url('application/json', configuration)
+    configuration = data_url('application/json', configuration),
+    port = { metrics = 9421 }, -- see https://github.com/prometheus/prometheus/wiki/Default-port-allocations
```
Comment (davidor) [inline gateway/config/development.lua]: :+1:

File: gateway/conf/nginx.conf.liquid (hunk 4 above)
Comment (mikz) [reply gateway/conf/nginx.conf.liquid]: It is used in https://github.com/3scale/apicast-cloud-hosted/pull/5/files#diff-047b1780b0ffeb4eba7b3d05beb76d5eR77 to extract the connection stats. 
There is no easy way to add that location from customization, so I added it here as this will become part of upstream at some point.

Comment (davidor) [reply gateway/conf/nginx.conf.liquid]: :+1:

File: gateway/src/apicast/policy.lua [hunk 6]
```diff
@@ -13,7 +13,7 @@ local PHASES = {
     'rewrite', 'access',
     'content', 'balancer',
     'header_filter', 'body_filter',
-    'post_action',  'log'
+    'post_action',  'log', 'metrics',
```
Comment (davidor) [inline gateway/src/apicast/policy.lua]: We should add new phases to the policies doc: https://github.com/3scale/apicast/blob/master/doc/policies.md#policies

File: gateway/conf/nginx.conf.liquid (hunk 3 above)
Comment (mikz) [reply gateway/conf/nginx.conf.liquid]: Done in https://github.com/3scale/apicast-cloud-hosted/pull/5/commits/09f65c93775c4bf6186aac8c92cdd06344e4f63d

File: gateway/src/apicast/executor.lua [hunk 7]
```diff
@@ -58,4 +59,11 @@ function _M:context(phase)
     return shared_build_context(self)
 end
 
+local metrics = _M.metrics
+--- Render metrics from all policies.
+function _M:metrics(...)
```
Comment (davidor) [inline gateway/src/apicast/executor.lua]: Maybe we should add a test for this. Define 2 policies with only a metrics phase that incr some counter and make sure that this returns the sum, or something similar.

File: gateway/src/apicast/policy.lua (hunk 6 above)
Comment (mikz) [reply gateway/src/apicast/policy.lua]: Done in https://github.com/3scale/apicast/pull/629/commits/c44b062f1047b718ac3a65773363708b1b1a704b

File: gateway/src/apicast/executor.lua (hunk 7 above)
Comment (mikz) [reply gateway/src/apicast/executor.lua]: Yeah but this does not really "return" anything. It renders the output.
I'll definitely do that when moving some policies with metrics to upstream from https://github.com/3scale/apicast-cloud-hosted/pull/5.

File: gateway/src/apicast/prometheus.lua [hunk 8]
```diff
@@ -0,0 +1,23 @@
+local prometheus = require('nginx.prometheus')
+local assert = assert
+local dict = 'prometheus_metrics'
+
+if ngx.shared[dict] then
+  local init = prometheus.init(dict)
+
+  local metrics = { }
+  local __call = function(_, type, name, ...)
+    local metric_name = assert(name, 'missing metric name')
```
Comment (davidor) [inline gateway/src/apicast/prometheus.lua]: Why do we check `name` and `type`? Can't we delegate this to the Prometheus lib?

File: spec/prometheus_spec.lua [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+      }
+    end)
+
+    local prometheus
+
+    before_each(function()
+      prometheus = assert(require('apicast.prometheus'))
+    end)
+
+    it('can be called', function()
```
Comment (davidor) [inline spec/prometheus_spec.lua]: Should we change this to test that it delegates `counter`, `gauge`, etc. to the Prometheus lib?

File: gateway/src/apicast/prometheus.lua (hunk 8 above)
Comment (mikz) [reply gateway/src/apicast/prometheus.lua]: Because we actually use it. The type gets a function that is later called and the name is used as a key to store the metric. I had few tricky errors so decided to go for explicit ones.

File: spec/prometheus_spec.lua (hunk 9 above)
Comment (mikz) [reply spec/prometheus_spec.lua]: Yep, makes sense.

Comment (mikz) [reply spec/prometheus_spec.lua]: Done.
